### ISSUE TITLE
vterm only redrawing on input

### ISSUE THREAD

**Issue by tazjin** [akermu/emacs-libvterm#555]
Since the last software update I did, vterm only redraws output when it receives more input, or when I perform something that causes Emacs to redraw buffers (e.g. changing the window layout).

Sometimes it does occasional updates on its own, but I am not sure what triggers them.

I've been poking around a bit and haven't found an obvious problem. Curiously, it seems that nothing changed in vterm *itself* between the set of packages where everything worked and the set where this occurs. I decided to open an issue here anyways in case somebody else has run into this and found a way to resolve it.

In the meantime, I'll attempt to bisect what is going on here.

Note: I use Emacs with native compilation.

**Comment by phaer** [comment]
I experience the same symptoms with Emacs 28.1. Did you find something by bisecting, @tazjin? Interestingly enough, I am on NixOS as well.

**Comment by tazjin** [comment]
Nope, never found the source of the issue and at some point after a subsequent update it just stopped v0v

For what it's worth, I'm on `nixos-unstable`

**Comment by phaer** [comment]
Ah, thanks for the quick response and good to know that there's a chance that it might be fixed in `nixos-22.05` (following 21.11 here)

**Comment by watanany** [comment]
Like @titibandit, I have a problem with Haskell's GHCi and `elm repl` not showing prompt or echoing input.

Here is the code that mimics [GHCi's line read code](https://github.com/ghc/ghc/blob/master/ghc/GHCi/UI.hs#L801), accepting input from the user and outputting it as is.
It doesn't show prompt and echo input from the user until a line feed is performed.

```haskell
import qualified System.Console.Haskeline as H

main :: IO ()
main = do
    -- putStr "\n"  -- (1)
    H.runInputT H.defaultSettings $ do
        maybeLine <- H.getInputLine "prompt >>> "
        case maybeLine of
          Just line -> H.outputStrLn line
          Nothing -> H.outputStrLn "No Input"
```

The following code also does not echo input from the user.

```haskell
main :: IO ()
main = do
    -- putStr "\n"  -- (2)
    line <- getLine
    print line
```

Uncommenting (1) and (2) in the above code makes it show prompt and makes it echo input from the user, but I don't know why.


This is the content of ~/.emacs.

```elisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)

(setq exec-path (append exec-path '("/opt/homebrew/bin")))
(use-package vterm :ensure t)
```

OS: macOS Monterey(12.4)
CPU: Apple M1

**Comment by titibandit** [comment]
I tried using old commits of this repository, and I can't fix the problem this way. It might have something to do with emacs itself.

**Comment by chaoky** [comment]
I experiencing the same issue, also on arch with the latest emacs/libvterm, but I'm also using doom

**Comment by jixiuf** [comment]
Does #617  fix this?  #598 has been fixed .
if it is not fixed, Could you tell me the step to reproduce your problem with a screenshot

**Comment by chaoky** [comment]
After most inputs, nothing will happen, even if I wait for a while, clicking or pressing some other key will then update the terminal with the last one.

After pressing a key for a while, the behaviour changes, that key will only be laggy, but It&rsquo;ll appear on it&rsquo;s own, it&rsquo;ll also become less laggy the more I press it.

I started having this problem after switching from [28.1.50 gcc+wayland](https://aur.archlinux.org/packages/emacs-gcc-wayland-devel-bin) to [28.1-7 native-comp](https://archlinux.org/packages/extra/x86_64/emacs-nativecomp/)

![image](https://user-images.githubusercontent.com/9826702/187016034-f679a39c-6504-40fe-a61e-7564fafa9708.png)

**Comment by titibandit** [comment]
I've updated vterm today, and the issue is fixed for me. I don't know if it's that pull request you mentioned but it's definitely fixed for me. Thanks!

**Comment by juhp** [comment]
Same - I can now use ghci inside a vterm :+1:
